runtime/query.go

```diff
@@ -45,362 +46,294 @@ func (*defaultQueryParser) Parse(msg proto.Message, values url.Values, filter *u
 		}
 	}
 	return nil
 }
 
 // PopulateFieldFromPath sets a value in a nested Protobuf structure.
-// It instantiates missing protobuf fields as it goes.
 func PopulateFieldFromPath(msg proto.Message, fieldPathString string, value string) error {
 	fieldPath := strings.Split(fieldPathString, ".")
-	return populateFieldValueFromPath(msg, fieldPath, []string{value})
+	return populateFieldValueFromPath(msg.ProtoReflect(), fieldPath, []string{value})
 }
 
-func populateFieldValueFromPath(msg proto.Message, fieldPath []string, values []string) error {
-	m := reflect.ValueOf(msg)
-	if m.Kind() != reflect.Ptr {
-		return fmt.Errorf("unexpected type %T: %v", msg, msg)
+func populateFieldValueFromPath(msgValue protoreflect.Message, fieldPath []string, values []string) error {
+	if len(fieldPath) < 1 {
+		return fmt.Errorf("no field path")
 	}
-	var props *proto.Properties
-	m = m.Elem()
+	if len(values) < 1 {
+		return fmt.Errorf("no value provided")
+	}
+
+	var fieldDesc protoreflect.FieldDescriptor
 	for i, fieldName := range fieldPath {
-		isLast := i == len(fieldPath)-1
-		if !isLast && m.Kind() != reflect.Struct {
-			return fmt.Errorf("non-aggregate type in the mid of path: %s", strings.Join(fieldPath, "."))
+		fields := msgValue.Descriptor().Fields()
+
+		// Get field by name
+		fieldDesc = fields.ByName(protoreflect.Name(fieldName))
+		if fieldDesc == nil {
+			fieldDesc = fields.ByJSONName(fieldName)
+			if fieldDesc == nil {
+				grpclog.Infof("field not found in %q: %q", msgValue.Descriptor().FullName(), strings.Join(fieldPath, "."))
+				return nil
+			}
 		}
-		var f reflect.Value
-		var err error
-		f, props, err = fieldByProtoName(m, fieldName)
-		if err != nil {
-			return err
-		} else if !f.IsValid() {
-			grpclog.Infof("field not found in %T: %s", msg, strings.Join(fieldPath, "."))
-			return nil
+
+		// If this is the last element, we're done
+		if i == len(fieldPath)-1 {
+			break
 		}
 
-		switch f.Kind() {
-		case reflect.Bool, reflect.Float32, reflect.Float64, reflect.Int32, reflect.Int64, reflect.String, reflect.Uint32, reflect.Uint64:
-			if !isLast {
-				return fmt.Errorf("unexpected nested field %s in %s", fieldPath[i+1], strings.Join(fieldPath[:i+1], "."))
-			}
-			m = f
-		case reflect.Slice:
-			if !isLast {
-				return fmt.Errorf("unexpected repeated field in %s", strings.Join(fieldPath, "."))
-			}
-			// Handle []byte
-			if f.Type().Elem().Kind() == reflect.Uint8 {
-				m = f
-				break
-			}
-			return populateRepeatedField(f, values, props)
-		case reflect.Ptr:
-			if f.IsNil() {
-				m = reflect.New(f.Type().Elem())
-				f.Set(m.Convert(f.Type()))
-			}
-			m = f.Elem()
-			continue
-		case reflect.Struct:
-			m = f
-			continue
-		case reflect.Map:
-			if !isLast {
-				return fmt.Errorf("unexpected nested field %s in %s", fieldPath[i+1], strings.Join(fieldPath[:i+1], "."))
-			}
-			return populateMapField(f, values, props)
+		switch fieldDesc.Kind() {
+		case protoreflect.MessageKind, protoreflect.GroupKind:
```

Comment:
```go
if fieldDesc.Message() != nil && field.Cardinality() != protoreflect.Repeated
```

It seems that the intent is to ensure that this is a *singular* message field. The current check will let lists of messages and maps through the cracks.

Reply:
Good spot! I added a test for this and it led me to find another bug in the code and the new code manages to get rid of passing the messageValue a lot, which is much nicer. Thanks!